File: src/authority.ts
import { q } from "./db";

export type AuthorityCredential = { id: string; publicKey: string };
export type AuthorityMember = {
  userId: string; role: "owner" | "operator" | "viewer"; credentials: AuthorityCredential[];
};
export type AuthorityApiKey = { id: string; userId: string; publicKey: string; scopes: string[]; expiresAt: number };
export type AuthorityMcpGrant = { id: string; userId: string; hash: string };
export type AuthoritySnapshot = {
  v: 1; workspaceId: string; members: AuthorityMember[]; apiKeys: AuthorityApiKey[]; mcpGrants?: AuthorityMcpGrant[];
};

export function authoritySnapshot(workspaceId: string): AuthoritySnapshot {
  const members = q<any>(`SELECT wm.user_id userId,wm.role,p.credential_id credentialId,p.public_key publicKey
    FROM workspace_members wm LEFT JOIN passkeys p ON p.user_id=wm.user_id
    WHERE wm.workspace_id=? ORDER BY wm.user_id,p.credential_id`).all(workspaceId);
  const grouped = new Map<string, AuthorityMember>();
  for (const row of members) {
    let member = grouped.get(row.userId);
    if (!member) {
      member = { userId: row.userId, role: row.role, credentials: [] };
      grouped.set(row.userId, member);
    }
    if (row.credentialId && row.publicKey) member.credentials.push({ id: row.credentialId, publicKey: row.publicKey });
  }
  const apiKeys = q<any>(`SELECT a.id,a.user_id userId,a.public_key publicKey,a.scopes,a.expires_at expiresAt
    FROM api_tokens a JOIN workspace_members wm ON wm.user_id=a.user_id
    WHERE wm.workspace_id=? AND a.public_key<>'' AND (a.expires_at=0 OR a.expires_at>?) ORDER BY a.id`).all(workspaceId, Date.now()).map(row => ({
      id: String(row.id), userId: String(row.userId), publicKey: String(row.publicKey),
      scopes: JSON.parse(String(row.scopes || "[]")) as string[], expiresAt: Number(row.expiresAt || 0),
    }));
  const mcpGrants = q<{ id: string; user_id: string; grant: string }>(`SELECT DISTINCT g.id,g.user_id,g.grant FROM mcp_grants g
    JOIN json_each(json_extract(g.grant,'$.deviceIds')) granted JOIN devices d ON d.id=granted.value
    WHERE d.workspace_id=? AND g.revoked_at IS NULL AND (g.expires_at=0 OR g.expires_at>?) AND EXISTS (
      SELECT 1 FROM json_each(json_extract(g.grant,'$.scopes')) scope WHERE scope.value IN ('mcp:actions','mcp:terminal')
    ) ORDER BY g.id`).all(workspaceId, Date.now())
    .map(row => ({ id: row.id, userId: row.user_id, hash: new Bun.CryptoHasher("sha256").update(row.grant).digest("hex") }));
  return { v: 1, workspaceId, members: [...grouped.values()], apiKeys, ...(mcpGrants.length ? { mcpGrants } : {}) };
}

export function canonicalAuthority(snapshot: AuthoritySnapshot) {
  return JSON.stringify(snapshot);
}

export function authorityHash(workspaceId: string) {
  const snapshot = canonicalAuthority(authoritySnapshot(workspaceId));
  return { snapshot, hash: new Bun.CryptoHasher("sha256").update(snapshot).digest("hex") };
}

File: crates/rc-server/src/authority.rs
use crate::{Database, now_ms, passkey_authority_material};
use rc_protocol::{
    AuthorityApiKey, AuthorityCredential, AuthorityMcpGrant, AuthorityMember, AuthoritySnapshot,
};
use sha2::{Digest, Sha256};
use std::collections::{BTreeMap, BTreeSet};
use webauthn_rs::prelude::Passkey;

pub fn authority_snapshot(db: &Database, workspace_id: &str) -> anyhow::Result<AuthoritySnapshot> {
    let members = authority_members(db, workspace_id)?;
    let api_keys = authority_api_keys(db, workspace_id)?;
    let mcp_grants = authority_mcp_grants(db, workspace_id)?;
    Ok(AuthoritySnapshot {
        v: 1,
        workspace_id: workspace_id.to_owned(),
        members,
        api_keys,
        mcp_grants,
    })
}

pub fn canonical_authority(db: &Database, workspace_id: &str) -> anyhow::Result<String> {
    Ok(serde_json::to_string(&authority_snapshot(
        db,
        workspace_id,
    )?)?)
}

pub fn authority_hash(snapshot: &str) -> String {
    Sha256::digest(snapshot.as_bytes())
        .iter()
        .map(|byte| format!("{byte:02x}"))
        .collect()
}

pub fn bootstrap_snapshot_for_device(
    db: &Database,
    device_id: &str,
) -> anyhow::Result<Option<String>> {
    let workspace = db.with_connection(|db| {
        use rusqlite::OptionalExtension;
        db.query_row(
            "SELECT workspace_id FROM devices WHERE id=?",
            [device_id],
            |row| row.get::<_, String>(0),
        )
        .optional()
    })?;
    workspace
        .map(|workspace_id| canonical_authority(db, &workspace_id))
        .transpose()
}

fn authority_members(db: &Database, workspace_id: &str) -> anyhow::Result<Vec<AuthorityMember>> {
    let rows = db.with_connection(|db| {
        let mut statement = db.prepare(
            "SELECT wm.user_id,wm.role,p.credential_json FROM workspace_members wm LEFT JOIN passkeys p ON p.user_id=wm.user_id WHERE wm.workspace_id=? ORDER BY wm.user_id,p.id",
        )?;
        statement
            .query_map([workspace_id], |row| {
                Ok((
                    row.get::<_, String>(0)?,
                    row.get::<_, String>(1)?,
                    row.get::<_, Option<String>>(2)?,
                ))
            })?
            .collect::<Result<Vec<_>, _>>()
    })?;
    let mut grouped = BTreeMap::<String, AuthorityMember>::new();
    for (user_id, role, credential_json) in rows {
        let member = grouped
            .entry(user_id.clone())
            .or_insert_with(|| AuthorityMember {
                user_id,
                role,
                credentials: Vec::new(),
            });
        if let Some(json) = credential_json {
            let passkey: Passkey = serde_json::from_str(&json)?;
            let (id, public_key) = passkey_authority_material(&passkey)?;
            member
                .credentials
                .push(AuthorityCredential { id, public_key });
        }
    }
    Ok(grouped.into_values().collect())
}

fn authority_api_keys(db: &Database, workspace_id: &str) -> anyhow::Result<Vec<AuthorityApiKey>> {
    let now = now_ms();
    Ok(db.with_connection(|db| {
        let mut statement = db.prepare(
            "SELECT c.id,c.user_id,c.public_key,c.scopes,c.expires_at FROM clients c JOIN workspace_members wm ON wm.user_id=c.user_id WHERE wm.workspace_id=? AND c.kind='api' AND c.public_key<>'' AND (c.expires_at=0 OR c.expires_at>?) ORDER BY c.id",
        )?;
        statement
            .query_map(rusqlite::params![workspace_id, now], |row| {
                let scopes: String = row.get(3)?;
                Ok(AuthorityApiKey {
                    id: row.get(0)?,
                    user_id: row.get(1)?,
                    public_key: row.get(2)?,
                    scopes: serde_json::from_str(&scopes).unwrap_or_default(),
                    expires_at: row.get(4)?,
                })
            })?
            .collect::<Result<Vec<_>, _>>()
    })?)
}

fn authority_mcp_grants(
    db: &Database,
    workspace_id: &str,
) -> anyhow::Result<Vec<AuthorityMcpGrant>> {
    let devices = db.with_connection(|db| {
        let mut statement =
            db.prepare("SELECT id FROM devices WHERE workspace_id=? ORDER BY id")?;
        statement
            .query_map([workspace_id], |row| row.get::<_, String>(0))?
            .collect::<Result<BTreeSet<_>, _>>()
    })?;
    if devices.is_empty() {
        return Ok(Vec::new());
    }
    let rows = db.with_connection(|db| {
        let mut statement = db.prepare(
            "SELECT id,user_id,grant FROM mcp_grants WHERE revoked_at IS NULL AND (expires_at=0 OR expires_at>?) ORDER BY id",
        )?;
        statement
            .query_map([now_ms()], |row| {
                Ok((
                    row.get::<_, String>(0)?,
                    row.get::<_, String>(1)?,
                    row.get::<_, String>(2)?,
                ))
            })?
            .collect::<Result<Vec<_>, _>>()
    })?;
    let mut grants = Vec::new();
    for (id, user_id, grant) in rows {
        let value: serde_json::Value = serde_json::from_str(&grant)?;
        let terminal = value
            .get("scopes")
            .and_then(|value| value.as_array())
            .is_some_and(|scopes| {
                scopes
                    .iter()
                    .any(|scope| scope.as_str() == Some("mcp:terminal"))
            });
        let applies = value
            .get("deviceIds")
            .and_then(|value| value.as_array())
            .is_some_and(|ids| {
                ids.iter()
                    .filter_map(|id| id.as_str())
                    .any(|id| devices.contains(id))
            });
        let owner = db.with_connection(|db| {
            use rusqlite::OptionalExtension;
            db.query_row(
                "SELECT 1 FROM workspace_members WHERE workspace_id=? AND user_id=? AND role='owner'",
                rusqlite::params![workspace_id, user_id],
                |_| Ok(()),
            )
            .optional()
            .map(|value| value.is_some())
        })?;
        if terminal && owner && applies {
            grants.push(AuthorityMcpGrant {
                id,
                user_id,
                hash: authority_hash(&grant),
            });
        }
    }
    Ok(grants)
}
